File: LibraryApp/app/src/main/java/com/example/libraryapp/BookAdapter.java
package com.example.libraryapp;

import android.graphics.Color;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.TextView;
import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;
import java.util.List;

public class BookAdapter extends RecyclerView.Adapter<BookAdapter.BookViewHolder> {

    private List<Book> bookList;
    private OnBookListener onBookListener;

    public BookAdapter(List<Book> bookList, OnBookListener onBookListener) {
        this.bookList = bookList;
        this.onBookListener = onBookListener;
    }

    @NonNull
    @Override
    public BookViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(parent.getContext()).inflate(R.layout.item_book, parent, false);
        return new BookViewHolder(view, onBookListener);
    }

    @Override
    public void onBindViewHolder(@NonNull BookViewHolder holder, int position) {
        Book book = bookList.get(position);
        holder.title.setText(book.getTitle());
        holder.author.setText(book.getAuthor());
        holder.status.setText(book.getStatus());

        // Dynamic Styling based on Status
        if ("Available".equalsIgnoreCase(book.getStatus())) {
            holder.status.setTextColor(Color.parseColor("#4CAF50")); // Green
            holder.btnReserve.setEnabled(true);
            holder.btnReserve.setText("RESERVE");
            holder.btnReserve.setBackgroundColor(Color.parseColor("#6200EE"));
        } else {
            holder.status.setTextColor(Color.RED);
            holder.btnReserve.setEnabled(false); // Disable button
            holder.btnReserve.setText("TAKEN");
            holder.btnReserve.setBackgroundColor(Color.GRAY);
        }
    }

    @Override
    public int getItemCount() {
        return bookList.size();
    }

    public class BookViewHolder extends RecyclerView.ViewHolder {
        TextView title, author, status;
        Button btnReserve;

        public BookViewHolder(@NonNull View itemView, OnBookListener listener) {
            super(itemView);
            title = itemView.findViewById(R.id.tvBookTitle);
            author = itemView.findViewById(R.id.tvBookAuthor);
            status = itemView.findViewById(R.id.tvBookStatus);
            btnReserve = itemView.findViewById(R.id.btnReserve);

            btnReserve.setOnClickListener(v -> {
                if (listener != null && getAdapterPosition() != RecyclerView.NO_POSITION) {
                    listener.onReserveClick(bookList.get(getAdapterPosition()));
                }
            });
        }
    }

    public interface OnBookListener {
        void onReserveClick(Book book);
    }
}

File: src/main/java/com/example/libraryapp20/BookAdapter.kt
package com.example.libraryapp20

import android.app.AlertDialog
import android.graphics.Color
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Button
import android.widget.TextView
import android.widget.Toast
import androidx.recyclerview.widget.RecyclerView

class BookAdapter(
    private val bookList: List<Book>,
    private val onBookListener: OnBookListener
) : RecyclerView.Adapter<BookAdapter.BookViewHolder>() {

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): BookViewHolder {
        val view = LayoutInflater.from(parent.context)
            .inflate(R.layout.item_book, parent, false)
        return BookViewHolder(view)
    }

    override fun onBindViewHolder(holder: BookViewHolder, position: Int) {
        val book = bookList[position]

        holder.title.text = book.title ?: ""
        holder.author.text = book.author ?: ""
        holder.status.text = book.status ?: ""

        val statusText = book.status?.trim()?.lowercase() ?: ""

        if (statusText == "available") {
            holder.status.setTextColor(Color.parseColor("#4CAF50")) // Green
            holder.btnReserve.isEnabled = true
            holder.btnReserve.text = "RESERVE"
            holder.btnReserve.setBackgroundColor(Color.parseColor("#6200EE"))
        } else {
            holder.status.setTextColor(Color.RED)
            holder.btnReserve.isEnabled = false
            holder.btnReserve.text = "TAKEN"
            holder.btnReserve.setBackgroundColor(Color.RED)
        }
    }

    override fun getItemCount(): Int = bookList.size

    inner class BookViewHolder(itemView: View) :
        RecyclerView.ViewHolder(itemView) {

        val title: TextView = itemView.findViewById(R.id.tvBookTitle)
        val author: TextView = itemView.findViewById(R.id.tvBookAuthor)
        val status: TextView = itemView.findViewById(R.id.tvBookStatus)
        val btnReserve: Button = itemView.findViewById(R.id.btnReserve)

        init {
            btnReserve.setOnClickListener {
                val position = bindingAdapterPosition
                if (position != RecyclerView.NO_POSITION) {
                    val book = bookList[position]

                    if (book.status?.lowercase() == "available") {

                        // 🔔 Confirmation Dialog
                        AlertDialog.Builder(itemView.context)
                            .setTitle("Confirm Reservation")
                            .setMessage("Are you sure you want to reserve \"${book.title}\"?")
                            .setPositiveButton("Yes") { _, _ ->
                                onBookListener.onReserveClick(book)
                            }
                            .setNegativeButton("Cancel", null)
                            .show()

                    } else {
                        Toast.makeText(
                            itemView.context,
                            "This book is not available.",
                            Toast.LENGTH_SHORT
                        ).show()
                    }
                }
            }
        }
    }

    interface OnBookListener {
        fun onReserveClick(book: Book)
    }
}
